src/ray/object_manager/object_directory.h

```diff
@@ -46,24 +46,39 @@ class ObjectDirectoryInterface {
                                      const InfoFailureCallback &fail_cb) = 0;
 
   /// Callback for object location notifications.
-  using OnLocationsFound = std::function<void(const std::vector<ray::ClientID> &v,
+  using OnLocationsFound = std::function<void(const std::vector<ray::ClientID> &,
                                               const ray::ObjectID &object_id)>;
 
+  /// Lookup object locations. Callback may be invoked with empty list of client ids.
+  ///
+  /// \param object_id The object's ObjectID.
+  /// \param callback Invoked with (possibly empty) list of client ids and object_id.
+  /// \return Status of whether async call to backend succeeded.
+  virtual ray::Status LookupLocations(const ObjectID &object_id,
+                                      const OnLocationsFound &callback) = 0;
+
   /// Subscribe to be notified of locations (ClientID) of the given object.
   /// The callback will be invoked whenever locations are obtained for the
   /// specified object.
   ///
+  /// \param callback_id The id associated with the specified callback. This is
+  /// needed when UnsubscribeObjectLocations is called.
   /// \param object_id The required object's ObjectID.
   /// \param success_cb Invoked with non-empty list of client ids and object_id.
   /// \return Status of whether subscription succeeded.
-  virtual ray::Status SubscribeObjectLocations(const ObjectID &object_id,
+  virtual ray::Status SubscribeObjectLocations(const UniqueID &callback_id,
```

Comment:
Can you add a `NOTE` to the documentation here that the callback may fire in the invocation of `SubscribeObjectLocations`? Until we can figure out a better way to do it, it'd be good to warn the user.